src/xwayland-unmanaged.c

```diff
@@ -35,6 +36,7 @@ unmanaged_handle_map(struct wl_listener *listener, void *data)
 	struct xwayland_unmanaged *unmanaged =
 		wl_container_of(listener, unmanaged, map);
 	struct wlr_xwayland_surface *xsurface = unmanaged->xwayland_surface;
+	assert(!unmanaged->node);
```

Comment:
I think these asserts() will allow a client application to crash the compositor by just sending two maps or unmaps in a row. Simply returning might be better instead.

We are also recreating the complete scene node based on the surface on an unmap / map cycle but even though that is not the best in case of latency I expect it not to be an issue in practice.

Reply:
It looks like wlroots already filters out duplicate map/unmap events; see:
https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/xwayland/xwm.c#L836
https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/xwayland/xwm.c#L853
https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/xwayland/xwm.c#L913

But I can add `return` statements instead of the `assert`s if we don't want to rely on the wlroots event filtering.

Reply:
Thanks for looking this up.

With that: Code LGTM, didn't test but feel free to merge.

Reply:
Thanks for the review!